aapenimservice/src/main/kotlin/no/nav/helsearbeidsgiver/inntektsmelding/aapenimservice/AapenImService.kt

```diff
@@ -0,0 +1,281 @@
+
+// TODO test
+class AapenImService(
+    private val rapid: RapidsConnection,
+    override val redisStore: RedisStore
+) : CompositeEventListener() {
+
+    private val logger = logger()
+    private val sikkerLogger = sikkerLogger()
+
+    override val event = EventName.AAPEN_IM_MOTTATT
+    override val startKeys = listOf(
+        Key.AAPEN_ID,
+        Key.SKJEMA_INNTEKTSMELDING,
+        Key.ARBEIDSGIVER_FNR
+    )
+    override val dataKeys = listOf(
+        Key.VIRKSOMHET,
+        Key.ARBEIDSTAKER_INFORMASJON,
+        Key.ARBEIDSGIVER_INFORMASJON,
+        Key.AAPEN_INNTEKTMELDING,
+        Key.ER_DUPLIKAT_IM
+    )
+
+    init {
+        StatefullEventListener(rapid, event, redisStore, startKeys, ::onPacket)
+        StatefullDataKanal(rapid, event, redisStore, dataKeys, ::onPacket)
+        FailKanal(rapid, event, ::onPacket)
+    }
+
+    override fun new(melding: Map<Key, JsonElement>) {
+        val transaksjonId = Key.UUID.les(UuidSerializer, melding)
+        val aapenId = Key.AAPEN_ID.les(UuidSerializer, melding)
+        val skjema = Key.SKJEMA_INNTEKTSMELDING.les(SkjemaInntektsmelding.serializer(), melding)
+        val avsenderFnr = Key.ARBEIDSGIVER_FNR.les(String.serializer(), melding)
+
+        MdcUtils.withLogFields(
+            Log.klasse(this),
+            Log.event(event),
+            Log.transaksjonId(transaksjonId),
+            Log.aapenId(aapenId)
+        ) {
+            rapid.publish(
+                Key.EVENT_NAME to event.toJson(),
+                Key.UUID to transaksjonId.toJson(),
+                Key.AAPEN_ID to aapenId.toJson(),
+                Key.BEHOV to BehovType.VIRKSOMHET.toJson(),
+                Key.ORGNRUNDERENHET to skjema.avsender.orgnr.toJson()
+            )
+
+            rapid.publish(
+                Key.EVENT_NAME to event.toJson(),
+                Key.UUID to transaksjonId.toJson(),
+                Key.AAPEN_ID to aapenId.toJson(),
+                Key.BEHOV to BehovType.FULLT_NAVN.toJson(),
+                Key.IDENTITETSNUMMER to skjema.sykmeldtFnr.toJson(),
+                Key.ARBEIDSGIVER_ID to avsenderFnr.toJson()
+            )
+        }
+    }
+
+    override fun inProgress(melding: Map<Key, JsonElement>) {
+        val transaksjonId = Key.UUID.les(UuidSerializer, melding)
+        val aapenId = Key.AAPEN_ID.les(UuidSerializer, melding)
+
+        MdcUtils.withLogFields(
+            Log.klasse(this),
+            Log.event(event),
+            Log.transaksjonId(transaksjonId),
+            Log.aapenId(aapenId)
+        ) {
+            if (isDataCollected(step1data(transaksjonId))) {
+                val skjema = redisStore.get(RedisKey.of(transaksjonId, Key.SKJEMA_INNTEKTSMELDING))?.fromJson(SkjemaInntektsmelding.serializer())
+                val orgNavn = redisStore.get(RedisKey.of(transaksjonId, Key.VIRKSOMHET)).orDefault(ukjentVirksomhet())
+
+                val sykmeldt = redisStore.get(RedisKey.of(transaksjonId, Key.ARBEIDSTAKER_INFORMASJON))
```

Comment:
lurer på om error-håndteringen gir deg tomPerson() i redis, og dermed blir det aldri brukt tomPerson(sykmeldtFnr)  ?

Reply:
Det stemmer. Dette er bare en "ekstra" for å slippe å bruke `!!`. Jeg har en ide om en generell forbedring. Kan se på det før denne merges.

Reply:
Nå er dette fikset, men det er avhengig av https://github.com/navikt/helsearbeidsgiver-inntektsmelding/pull/402.